nomad/blocked_evals.go

```diff
@@ -0,0 +1,303 @@
+package nomad
+
+import (
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// unblockBuffer is the buffer size for the unblock channel.
+	unblockBuffer = 8096
+)
+
+// BlockedEvals is used to track evaluations that shouldn't be queued until a
+// certain class of nodes becomes available. An evaluation is put into the
+// blocked state when it is run through the scheduler and produced failed
+// allocations. It is unblocked when the capacity of a node that could run the
+// failed allocation becomes available.
+type BlockedEvals struct {
+	evalBroker *EvalBroker
+	enabled    bool
+	running    bool
+	stats      *BlockedStats
+	l          sync.RWMutex
+
+	// captured is the set of evaluations that are captured by computed node
+	// classes.
+	captured map[string]*structs.Evaluation
+
+	// escaped is the set of evaluations that have escaped computed node
+	// classes.
+	escaped map[string]*structs.Evaluation
+
+	// unblockCh is used to buffer unblocking of evaluations.
+	unblockCh chan string
+
+	// jobs is the map of blocked job and is used to ensure that only one
+	// blocked eval exists for each job.
+	jobs map[string]struct{}
+
+	// duplicates is the set of evaluations for jobs that had pre-existing
+	// blocked evaluations. These should be marked as cancelled since only one
+	// blocked eval is neeeded bper job.
+	duplicates []*structs.Evaluation
+
+	// duplicateCh is used to signal that a duplicate eval was added to the
+	// duplicate set. It can be used to unblock waiting callers looking for
+	// duplicates.
+	duplicateCh chan struct{}
+
+	// stopCh is used to stop any created goroutines.
+	stopCh chan struct{}
+}
+
+// BlockedStats returns all the stats about the blocked eval tracker.
+type BlockedStats struct {
+	// TotalEscaped is the total number of blocked evaluations that have escaped
+	// computed node classes.
+	TotalEscaped int
+
+	// TotalBlocked is the total number of blocked evaluations.
+	TotalBlocked int
+}
+
+// NewBlockedEvals creates a new blocked eval tracker that will enqueue
+// unblocked evals into the passed broker.
+func NewBlockedEvals(evalBroker *EvalBroker) *BlockedEvals {
+	return &BlockedEvals{
+		evalBroker:  evalBroker,
+		captured:    make(map[string]*structs.Evaluation),
+		escaped:     make(map[string]*structs.Evaluation),
+		jobs:        make(map[string]struct{}),
+		unblockCh:   make(chan string, unblockBuffer),
+		duplicateCh: make(chan struct{}),
+		stopCh:      make(chan struct{}),
+		stats:       new(BlockedStats),
+	}
+}
+
+// Enabled is used to check if the broker is enabled.
+func (b *BlockedEvals) Enabled() bool {
+	b.l.RLock()
+	defer b.l.RUnlock()
+	return b.enabled
+}
+
+// SetEnabled is used to control if the broker is enabled. The broker
+// should only be enabled on the active leader.
+func (b *BlockedEvals) SetEnabled(enabled bool) {
+	b.l.Lock()
+	b.enabled = enabled
+	if !b.running {
```

Comment:
Right, but you can just check if enabled is already true in that case. E.g. if enabled == running in all cases, so no need to have both.